### PR TITLE
Improve and simplify maintenance of APF bootstrap objects without type assertions

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Five things:

- Remove unnecessary layers of indirection to reach constant logic.
- Remove unnecessary split between finding unwanted objects and removing them.
- Use interfaces to remove need for type assertions.
- Consistently use a context.Context tied to the server shutdown.
- This PR does _not_ make deletion of unwanted objects conditional on ResourceVersion.  That would be a functional improvement, which I am reserving for a follow-on PR.  However, this PR includes some TODOs highlighting where that follow-on will make changes.

#### Which issue(s) this PR fixes:

This PR does not address point 4 of #107097 but _does_ include TODOs about how to fix that in a follow-on PR.

#### Special notes for your reviewer:
This is #107704 brought back to life.
This is an alternative to #107696 .
The version has the advantage of not using type assertions (outside the tests).
Accomplishing this required carefully wrapping up data and operations in interfaces,
arguably making more convoluted code.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
The force-push to 6545605e010 is a rebase onto `master`.

**Comment by MikeSpreitzer** [comment]
pull-kubernetes-verify-govet-levee is failing everything now (#111452)

**Comment by fedebongio** [comment]
@tkashem we discussed this issue, you seem to be the only person available to review it among all the potential candidates... sorry to bug you!

cc @deads2k in case you can too

**Comment by MikeSpreitzer** [comment]
The force-push to 6f0cf1b29fb is a rebase onto master.

**Comment by MikeSpreitzer** [comment]
The force-push to cda6dc8d84e is a rebase onto `master`.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by MikeSpreitzer** [comment]
@dims : huh?  This is not marked as needing rebase.

**Comment by dims** [comment]
> @dims : huh? This is not marked as needing rebase.

my bad, yep you are good. (typo in my query caught this PR in the net)

**Review by tkashem — COMMENTED** [review]
this overall looks good to me - i see some behavior changes, can we limit this PR to the refactor only, and address the behavior changes in a new PR? Thanks!

**Comment by MikeSpreitzer** [comment]
@tkashem : can you please say what behavior differences you see?  I thought I already removed any, but it has been a while, so I might have forgotten something.

File: pkg/registry/flowcontrol/ensurer/strategy.go [hunk 1]
```diff
@@ -215,130 +208,132 @@ func setAutoUpdateAnnotation(accessor metav1.Object, autoUpdate bool) {
...
+			klog.V(5).InfoS("Skipping deletion of APF object with "+flowcontrolv1beta3.AutoUpdateAnnotationKey+"=false annotation", "name", name)
+			continue
+		}
+		// TODO: expectedResourceVersion := object.GetResourceVersion()
+		err = object.delete(ctx /* TODO: expectedResourceVersion */)
+		if err == nil {
+			klog.V(2).InfoS(fmt.Sprintf("Successfully deleted the unwanted %s", boots.typeName()), "name", name)
+			continue
 		}
+		if apierrors.IsConflict(err) {
```
Comment (tkashem) [inline pkg/registry/flowcontrol/ensurer/strategy.go:320]: if i am not mistaken, this seems unrelated to the refactor, and the context wiring also not related to the refactor, if not a separate PR, can we have three separate commits in this pr? thanks!

**Comment by MikeSpreitzer** [comment]
The force-push to a40e2eb75ae factors into two commits, one that re-organizes the code and a following commit that threads the context.

File: pkg/registry/flowcontrol/rest/storage_flowcontrol.go [hunk 2]
```diff
@@ -202,59 +209,43 @@ func ensure(clientset flowcontrolclient.FlowcontrolV1beta3Interface, fsLister fl
 }
 
 func ensureSuggestedConfiguration(clientset flowcontrolclient.FlowcontrolV1beta3Interface, fsLister flowcontrollisters.FlowSchemaLister, plcLister flowcontrollisters.PriorityLevelConfigurationLister) error {
-	plEnsurer := ensurer.NewSuggestedPriorityLevelEnsurerEnsurer(clientset.PriorityLevelConfigurations(), plcLister)
-	if err := plEnsurer.Ensure(flowcontrolbootstrap.SuggestedPriorityLevelConfigurations); err != nil {
+	plcSuggesteds := ensurer.WrapBootstrapPriorityLevelConfigurations(clientset.PriorityLevelConfigurations(), plcLister, flowcontrolbootstrap.SuggestedPriorityLevelConfigurations)
```
Comment (tkashem) [inline pkg/registry/flowcontrol/rest/storage_flowcontrol.go:214]: is it possible to merge the suggested PLs and flowshemas into one slice, so we have one `EnsureConfigurations` call for all the sugested?

File: pkg/registry/flowcontrol/rest/storage_flowcontrol.go [hunk 3]
```diff
@@ -202,59 +209,43 @@ func ensure(clientset flowcontrolclient.FlowcontrolV1beta3Interface, fsLister fl
...
-	fsEnsurer := ensurer.NewSuggestedFlowSchemaEnsurer(clientset.FlowSchemas(), fsLister)
-	return fsEnsurer.Ensure(flowcontrolbootstrap.SuggestedFlowSchemas)
+	fsSuggesteds := ensurer.WrapBootstrapFlowSchemas(clientset.FlowSchemas(), fsLister, flowcontrolbootstrap.SuggestedFlowSchemas)
+	return ensurer.EnsureConfigurations(fsSuggesteds, ensurer.NewSuggestedEnsureStrategy())
 }
 
 func ensureMandatoryConfiguration(clientset flowcontrolclient.FlowcontrolV1beta3Interface, fsLister flowcontrollisters.FlowSchemaLister, plcLister flowcontrollisters.PriorityLevelConfigurationLister) error {
-	fsEnsurer := ensurer.NewMandatoryFlowSchemaEnsurer(clientset.FlowSchemas(), fsLister)
-	if err := fsEnsurer.Ensure(flowcontrolbootstrap.MandatoryFlowSchemas); err != nil {
+	plcMandatories := ensurer.WrapBootstrapPriorityLevelConfigurations(clientset.PriorityLevelConfigurations(), plcLister, flowcontrolbootstrap.MandatoryPriorityLevelConfigurations)
```
Comment (tkashem) [inline pkg/registry/flowcontrol/rest/storage_flowcontrol.go:224]: same here, one slice for all the mandatory PLs and flowschemas

**Review by tkashem — COMMENTED** [review]
I have a few minor comments, this overall LGTM

File: pkg/registry/flowcontrol/rest/storage_flowcontrol.go (hunk 2 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/rest/storage_flowcontrol.go:214]: That would be no small change.  Remember that `ensurer.WrapBootstrapPriorityLevelConfigurations` is not returning a slice but rather an interface with the following signature.

```
// BootstrapObjects is a generic interface to a list of bootstrap objects bound up with the relevant operations on them.
// The binding makes it unnecessary to have any type casts.
// A bootstrap object is a mandatory or suggested config object,
// with the spec that the code is built to provide.
type BootstrapObjects interface {
	typeName() string                         // the Kind of the objects
	len() int                                 // number of objects
	get(int) bootstrapObject                  // extract one object, origin 0
	getExistingObjects() ([]deletable, error) // returns all the APF config objects that exist at the moment
}
```

and `RemoveUnwantedObjects` assumes that they all have distinct names.

File: pkg/registry/flowcontrol/ensurer/flowschema.go [hunk 4]
```diff
@@ -18,191 +18,139 @@ package ensurer
...
-
-	return getDanglingBootstrapObjectNames(bootstrapNames, currentObjects), nil
+	return dels, nil
 }
 
-type fsEnsurer struct {
-	strategy ensureStrategy
-	wrapper  configurationWrapper
+type bootstrapFlowSchema struct {
+	*bootstrapFlowSchemas
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/flowschema.go]: You're right, we do not.  The latest push factors `bootstrapFlowSchemas` so that the list is not carried where it is not needed.

**Comment by tkashem** [comment]
/lgtm
Thanks!

File: pkg/registry/flowcontrol/ensurer/flowschema.go [hunk 5]
```diff
@@ -18,191 +18,144 @@ package ensurer
...
-// FlowSchemaRemover is the interface that wraps the
-// RemoveAutoUpdateEnabledObjects method.
-//
-// RemoveAutoUpdateEnabledObjects removes a set of bootstrap FlowSchema
-// objects specified via their names. The function removes an object
-// only if automatic update of the spec is enabled for it.
-type FlowSchemaRemover interface {
-	RemoveAutoUpdateEnabledObjects([]string) error
+type bootstrapFlowSchemas struct {
+	flowSchemaClient
```
Comment (wojtek-t) [inline pkg/registry/flowcontrol/ensurer/flowschema.go:48]: nit: can we make it a named field:
```
client flowSchemaClient
```
?

File: pkg/registry/flowcontrol/ensurer/flowschema.go [hunk 6]
```diff
@@ -18,191 +18,144 @@ package ensurer
...
-
-	return getDanglingBootstrapObjectNames(bootstrapNames, currentObjects), nil
+	return dels, nil
 }
 
-type fsEnsurer struct {
-	strategy ensureStrategy
-	wrapper  configurationWrapper
+type bootstrapFlowSchema struct {
+	*flowSchemaClient
```
Comment (wojtek-t) [inline pkg/registry/flowcontrol/ensurer/flowschema.go:83]: same here?

File: pkg/registry/flowcontrol/ensurer/flowschema.go [hunk 7]
```diff
@@ -18,191 +18,144 @@ package ensurer
...
 }
 
-func (e *fsEnsurer) RemoveAutoUpdateEnabledObjects(flowSchemas []string) error {
-	for _, flowSchema := range flowSchemas {
-		if err := removeAutoUpdateEnabledConfiguration(e.wrapper, flowSchema); err != nil {
-			return err
-		}
+func (boot *bootstrapFlowSchema) getCurrent() (wantAndHave, error) {
+	current, err := boot.lister.Get(boot.bootstrap.Name)
+	if err != nil {
```
Comment (wojtek-t) [inline pkg/registry/flowcontrol/ensurer/flowschema.go:98]: Don't we want to handle `NotFound` error differently here?

File: pkg/registry/flowcontrol/ensurer/prioritylevelconfiguration.go [hunk 8]
```diff
@@ -18,192 +18,144 @@ package ensurer
...
-// RemoveAutoUpdateEnabledObjects method.
-//
-// RemoveAutoUpdateEnabledObjects removes a set of bootstrap
-// PriorityLevelConfiguration objects specified via their names.
-// The function removes an object only if automatic update
-// of the spec is enabled for it.
-type PriorityLevelRemover interface {
-	RemoveAutoUpdateEnabledObjects([]string) error
+type bootstrapPriorityLevelConfigurations struct {
+	priorityLevelConfigurationClient
```
Comment (wojtek-t) [inline pkg/registry/flowcontrol/ensurer/prioritylevelconfiguration.go:48]: nit: can we make it a named field:

File: pkg/registry/flowcontrol/ensurer/strategy.go [hunk 9]
```diff
@@ -215,130 +208,130 @@ func setAutoUpdateAnnotation(accessor metav1.Object, autoUpdate bool) {
...
-		klog.ErrorS(err, fmt.Sprintf("Skipping deletion of the %s", wrapper.TypeName()), "name", name)
-
-		// This may need manual intervention, in case the annotation value is malformed,
-		// so don't return an error, that might trigger futile retry loop.
-		return nil
-	}
-	if !autoUpdate {
-		klog.V(5).InfoS(fmt.Sprintf("Skipping deletion of the %s", wrapper.TypeName()), "name", name)
-		return nil
+		klog.ErrorS(err, "Failed to list existing APF configuration objects", "type", boots.typeName())
```
Comment (wojtek-t) [inline pkg/registry/flowcontrol/ensurer/strategy.go:286]: Shouldn't we handle NotFound error differently, as before?

File: pkg/registry/flowcontrol/rest/storage_flowcontrol.go [hunk 10]
```diff
@@ -282,3 +275,27 @@ func contextFromChannelAndMaxWaitDuration(stopCh <-chan struct{}, maxWait time.D
 	}()
 	return ctx, cancel
 }
+
+// contextFromChannel returns a Context that is bound to the
+// specified channel. The derived context will be
+// cancelled when the specified channel stopCh is closed.
+//
+// Note the caller must *always* call the CancelFunc, otherwise resources may be leaked.
+func contextFromChannel(stopCh <-chan struct{}) (context.Context, context.CancelFunc) {
```
Comment (wojtek-t) [inline pkg/registry/flowcontrol/rest/storage_flowcontrol.go:284]: There is already `ContextForChannel` in:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L116

Doesn't it do the same?

**Review by wojtek-t — COMMENTED** [review]
Just few minor comments - other than that it LGTM.

File: pkg/registry/flowcontrol/rest/storage_flowcontrol.go (hunk 10 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/rest/storage_flowcontrol.go:284]: Hmm, this is a copy of what was there when this was written; I do not remember why.

BTW, wait.ContextForChannel has been modified since, I am going to have to rebase.

File: pkg/registry/flowcontrol/ensurer/flowschema.go (hunk 7 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/flowschema.go:98]: The caller handles the NotFound error.  The `wantAndHave` interface does not have a decent interpretation in the NotFound case.

File: pkg/registry/flowcontrol/ensurer/strategy.go (hunk 9 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/strategy.go:286]: (1) Does a lister ever return a NotFound error?
(2) If the lister ever actually did return a NotFound error, there is no more work to do about removing unwanted objects.
(3) I think the behavior here matches the behavior of what is currently merged.

**Comment by MikeSpreitzer** [comment]
The force-push to cdcf10f1a77 is a rebase plus one more commit that responds to the latest review.

File: pkg/registry/flowcontrol/ensurer/flowschema.go (hunk 5 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/flowschema.go:48]: We could but what would be the benefit? The costs would include writing more methods because of promotions that are no longer there.

File: pkg/registry/flowcontrol/ensurer/flowschema.go (hunk 6 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/flowschema.go:83]: We could but what would be the benefit? The costs would include writing more methods because of promotions that are no longer there.

File: pkg/registry/flowcontrol/ensurer/prioritylevelconfiguration.go (hunk 8 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/prioritylevelconfiguration.go:48]: We could but what would be the benefit? The costs would include writing more methods because of promotions that are no longer there.

File: pkg/registry/flowcontrol/ensurer/strategy.go (hunk 9 above)
Comment (wojtek-t) [reply pkg/registry/flowcontrol/ensurer/strategy.go:286]: > (1) Does a lister ever return a NotFound error?

If the object is not in the cache, it will return NotFound.
But I agree not for list.

> (3) I think the behavior here matches the behavior of what is currently merged.

You're proceeding with further processing in case of non-nil error.
But in that case, current might be nil.
This should work fine, but I would feel safer if we simply return here.

**Review by wojtek-t — COMMENTED** [review]
Just one minor small nit - other than that LGTM.

File: pkg/registry/flowcontrol/ensurer/strategy.go (hunk 9 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/strategy.go]: Oh, I see.  Actually, on further examination, I see that returning an error in this case is what matches the current behavior.  Fixed in the latest revision.

**Comment by MikeSpreitzer** [comment]
The latest revision also adds bootstrap object copying, analogous to #117107 , along with comments both asserting immutability and explaining that the Create call may side-effect.

File: pkg/registry/flowcontrol/ensurer/strategy.go [hunk 11]
```diff
@@ -215,130 +208,130 @@ func setAutoUpdateAnnotation(accessor metav1.Object, autoUpdate bool) {
...
-// removeAutoUpdateEnabledConfiguration makes an attempt to remove the given
-// configuration object if automatic update of the spec is enabled for this object.
-func removeAutoUpdateEnabledConfiguration(wrapper configurationWrapper, name string) error {
-	current, err := wrapper.Get(name)
+// RemoveUnwantedObjects attempts to delete the configuration objects
+// that exist, are annotated `apf.kubernetes.io/autoupdate-spec=true`, and do not
+// have a name in the given set.  A refusal due to concurrent update is logged
+// and not considered an error; the object will be reconsidered later.
+func RemoveUnwantedObjects(ctx context.Context, boots BootstrapObjects) error {
+	current, err := boots.getExistingObjects()
```
Comment (wojtek-t) [inline pkg/registry/flowcontrol/ensurer/strategy.go:284]: Hmm - why you return nil here?
I would suggest returning the same error that we were returning before:
```
return fmt.Errorf("failed to retrieve the %s, will retry later name=%q error=%w", wrapper.TypeName(), name, err)
```

This error is bubbles up to `ensure()` method so it matter in two places:
https://github.com/kubernetes/kubernetes/pull/111422/files#diff-2db4d80439afc0612e686dfe2939c42ff1e620997a123d4fd5453ecbf7cec009R158
https://github.com/kubernetes/kubernetes/pull/111422/files#diff-2db4d80439afc0612e686dfe2939c42ff1e620997a123d4fd5453ecbf7cec009R180

In the first one, I would say that retrying is desired.
In the second, it only adds an error log, which also seems desired.

**Review by wojtek-t — COMMENTED** [review]
I'm sorry - just one more comment about the error.

File: pkg/registry/flowcontrol/ensurer/strategy.go (hunk 11 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/strategy.go:284]: Oh crud!  That's a typo.  I meant to `return err`.  Fixed in the latest rev.

**Comment by MikeSpreitzer** [comment]
The force-push to 008576da073 is a squash and rebase; the rebase is required by the conflict introduced by the concurrent merge of #117107 .

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!
